src/server/pps/server/datum_set.go

```diff
@@ -1,94 +1,133 @@
 package server
 
 import (
+	"fmt"
+
 	"github.com/pachyderm/pachyderm/src/client/pfs"
 	"github.com/pachyderm/pachyderm/src/client/pps"
+	workerpkg "github.com/pachyderm/pachyderm/src/server/pkg/worker"
 
 	"golang.org/x/net/context"
 )
 
 type datumFactory interface {
-	Next() []*pfs.FileInfo
-	// Resets the internal indexes so we start reading from the first
-	// datum set again.
-	Reset()
-	Indexes() []int
 	Len() int
+	Datum(i int) []*workerpkg.Input
 }
 
-type datumFactoryImpl struct {
-	indexes    []int
-	datumLists [][]*pfs.FileInfo
-	done       bool
+type atomDatumFactory struct {
+	inputs []*workerpkg.Input
+	index  int
 }
 
-func (d *datumFactoryImpl) Next() []*pfs.FileInfo {
-	if d.done {
-		return nil
+func newAtomDatumFactory(ctx context.Context, pfsClient pfs.APIClient, input *pps.AtomInput) (datumFactory, error) {
+	result := &atomDatumFactory{}
+	fileInfos, err := pfsClient.GlobFile(ctx, &pfs.GlobFileRequest{
+		Commit:  input.Commit,
+		Pattern: input.Glob,
+	})
+	if err != nil {
+		return nil, err
+	}
+	for _, fileInfo := range fileInfos.FileInfo {
+		result.inputs = append(result.inputs, &workerpkg.Input{
+			FileInfo: fileInfo,
+			Name:     input.Name,
+			Lazy:     input.Lazy,
+		})
 	}
+	return result, nil
+}
 
-	defer func() {
-		// Increment the indexes
-		for i := 0; i < len(d.datumLists); i++ {
-			if d.indexes[i] == len(d.datumLists[i])-1 {
-				d.indexes[i] = 0
-				continue
-			}
-			d.indexes[i]++
-			return
-		}
-		d.done = true
-	}()
+func (d *atomDatumFactory) Len() int {
+	return len(d.inputs)
+}
+
+func (d *atomDatumFactory) Datum(i int) []*workerpkg.Input {
+	return []*workerpkg.Input{d.inputs[i]}
+}
+
+type unionDatumFactory struct {
+	inputs []datumFactory
+}
 
-	var datum []*pfs.FileInfo
-	for i, index := range d.indexes {
-		datum = append(datum, d.datumLists[i][index])
+func newUnionDatumFactory(ctx context.Context, pfsClient pfs.APIClient, unionInput *pps.UnionInput) (datumFactory, error) {
+	result := &unionDatumFactory{}
+	for _, input := range unionInput.Input {
+		datumFactory, err := newDatumFactory(ctx, pfsClient, input)
+		if err != nil {
+			return nil, err
+		}
+		result.inputs = append(result.inputs, datumFactory)
 	}
-	return datum
+	return result, nil
 }
 
-func (d *datumFactoryImpl) Indexes() []int {
-	return d.indexes
+func (d *unionDatumFactory) Len() int {
+	result := 0
+	for _, datumFactory := range d.inputs {
+		result += datumFactory.Len()
+	}
+	return result
 }
 
-func (d *datumFactoryImpl) Reset() {
-	for i := range d.indexes {
-		d.indexes[i] = 0
+func (d *unionDatumFactory) Datum(i int) []*workerpkg.Input {
+	for _, datumFactory := range d.inputs {
+		if i < datumFactory.Len() {
+			return datumFactory.Datum(i)
+		}
+		i -= datumFactory.Len()
 	}
+	panic("index out of bounds")
```

Comment:
Do we want to `panic`?  I think panicing is fine if this is a truly exceptional, unrecoverable case, but just wanted to double check.